runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -301,62 +271,245 @@ pub(crate) fn wasmer2_vm_hash() -> u64 {
+        let compiler = Singlepass::new();
+        // We only support universal engine at the moment.
+        assert_eq!(WASMER2_CONFIG.engine, WasmerEngine::Universal);
+        Self { config, engine: Universal::new(compiler).features(WASMER_FEATURES).engine() }
+    }
+
+    pub(crate) fn compile_uncached(&self, code: &[u8]) -> Result<VMArtifact, CompilationError> {
+        self.engine
+            .validate(code)
+            .map_err(|e| CompilationError::WasmerCompileError { msg: e.to_string() })?;
+        Ok(VMArtifact {
+            artifact: self
+                .engine
+                .compile(code, &self)
+                .map_err(|e| CompilationError::WasmerCompileError { msg: e.to_string() })?,
+            _engine: self.engine.clone(),
+        })
+    }
+
+    pub(crate) unsafe fn deserialize(
+        &self,
+        serialized: &[u8],
+    ) -> Result<VMArtifact, DeserializeError> {
+        Ok(VMArtifact {
+            artifact: self.engine.deserialize(serialized)?,
+            _engine: self.engine.clone(),
+        })
+    }
+
+    fn run_method(
+        &self,
+        artifact: &VMArtifact,
+        mut import: Wasmer2Imports<'_, '_>,
+        method_name: &str,
+    ) -> Result<(), VMError> {
+        let _span = tracing::debug_span!(target: "vm", "run_method").entered();
+
+        // FastGasCounter in Nearcore and Wasmer must match in layout.
+        assert_eq!(size_of::<FastGasCounter>(), size_of::<wasmer_types::FastGasCounter>());
+        assert_eq!(
+            offset_of!(FastGasCounter, burnt_gas),
+            offset_of!(wasmer_types::FastGasCounter, burnt_gas)
+        );
+        assert_eq!(
+            offset_of!(FastGasCounter, gas_limit),
+            offset_of!(wasmer_types::FastGasCounter, gas_limit)
+        );
+        assert_eq!(
+            offset_of!(FastGasCounter, opcode_cost),
+            offset_of!(wasmer_types::FastGasCounter, opcode_cost)
+        );
+        let gas = import.vmlogic.gas_counter_pointer() as *mut wasmer_types::FastGasCounter;
+
+        let entrypoint = get_entrypoint_index(&*artifact.artifact, method_name)?;
+        unsafe {
+            let instance = {
+                let _span = tracing::debug_span!(target: "vm", "run_method/instantiate").entered();
+                // SAFETY/FIXME: this transmute shouldn't be necessary, but we are limited by
+                // wasmer's API right now.
+                //
+                // An important caveat is that the `'static` lifetime here refers to the lifetime
+                // of `VMLogic` reference to which is retained by the `InstanceHandle` we create.
+                // However this `InstanceHandle` only lives during the execution of this body, so
+                // we can be sure that `VMLogic` remains live and valid at any time.
+                let import =
+                    std::mem::transmute::<_, &mut Wasmer2Imports<'static, 'static>>(&mut import);
```

Comment:
```suggestion
                    std::mem::transmute::<&mut Wasmer2Imports<_, _>, &mut Wasmer2Imports<'static, 'static>>(&mut import);
```

just to be extra-paranoid that we convert reference of the right type.